### DISCUSSION TITLE
I am getting error after adding custom middleware for filament authentication

### DISCUSSION THREAD

**Discussion by mohandtharsha** [filamentphp/filament#9367] [Help]
I have created custom middleware for filament authentication to restrict users based on roles and active status But I am getting the below error,Filament\FilamentManager::getUserAvatarUrl(): Argument #1 ($user) must be of type Illuminate\Database\Eloquent\Model|Illuminate\Contracts\Auth\Authenticatable, null given, called in C:\xampp\htdocs\EldersTalentNew\storage\framework\views\96bb8307b3bfe33a8c5d0701809c42c5.php on line 22

Below is my code,

<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;
use Symfony\Component\HttpFoundation\Response;
use App\Models\Roles;
use Illuminate\Validation\ValidationException;
use App\Exceptions\PermissionDeniedException;
use Illuminate\Support\Facades\Redirect;
use Illuminate\Support\Facades\Session;
use Filament\Facades\Filament;
use Filament\Notifications\Notification;

class CheckUserRole
{
    public function handle(Request $request, Closure $next, $role)
    {
        if (Auth::check()) {
            $roleName = ltrim($role, ':');
            $authUser = Auth::user();

            if (!empty($authUser)) {
                $role_id = $authUser->role_id;
                $status = $authUser->active_status;
                $role = Roles::where('id', $role_id)->first();
                $role_name = $role->name;

                if ($roleName === $role_name) {
                    if ($status === 0) {
                        // No need to call Auth::login() if the user is already logged in.
                        // Just return the response and continue with the request.
                        return $next($request);
                    } else {
                        Auth::logout();
                        Notification::make()
                            ->title('Your account is deactivated!')
                            ->danger()
                            ->send();
                    }
                } else {
                    Auth::logout();
                    Notification::make()
                        ->title('You do not have permission to log in to this account!')
                        ->danger()
                        ->send();
                }
            }
        }
        
        return $next($request);
    }

}

**Comment by VincvanAs** [comment]
Could you share how you define the  $authUser?

You could also use the canAccessPanel in your user model to arrange this.

  ```
public function canAccessPanel(Panel $panel): bool
    {
return str_ends_with($this->email, '@example.com') && $this->active_status;
    }
```

**Reply by mohandtharsha** [reply]
public function handle(Request $request, Closure $next, $role)
    {
        if (Auth::check()) {
            $roleName = ltrim($role, ':');
            $authUser = Auth::user();

            if (!empty($authUser)) {
                $role_id = $authUser->role_id;
                $status = $authUser->active_status;
                $role = Roles::where('id', $role_id)->first();
                $role_name = $role->name;

                if ($roleName === $role_name) {
                    if ($status === 0) {
                        // No need to call Auth::login() if the user is already logged in.
                        // Just return the response and continue with the request.
                        return $next($request);
                    } else {
                        Auth::logout();
                        Notification::make()
                            ->title('Your account is deactivated!')
                            ->danger()
                            ->send();
                    }
                } else {
                    Auth::logout();
                    Notification::make()
                        ->title('You do not have permission to log in to this account!')
                        ->danger()
                        ->send();
                }
            }
        } Please check this I want to do it here only

**Reply by VincvanAs** [reply]
Did you defined HasAvatar in your user model? And in addition

   ` public function getFilamentAvatarUrl(): ?string
    {
        return $this->avatar_url;
    }`

Looks like something wrong with your users avatar

**Reply by mohandtharsha** [reply]
I want to logout using filament 

<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;
use Symfony\Component\HttpFoundation\Response;
use App\Models\Roles;
use Illuminate\Validation\ValidationException;
use App\Exceptions\PermissionDeniedException;
use Illuminate\Support\Facades\Redirect;
use Illuminate\Support\Facades\Session;
use Filament\Notifications\Notification;

class CheckUserRole
{
    public function handle(Request $request, Closure $next, $role)
    {
        if (Auth::check()) {
            $roleName = ltrim($role, ':');
            $authUser = Auth::user();
            if ($authUser !== null) { // Check if a user is logged in and the user object is not null.
                $role_id = $authUser->role_id;
                $status = $authUser->active_status;
                $role = Roles::where('id', $role_id)->first();
                $role_name = $role->name;

                if ($roleName === $role_name) {
                    if ($status === 0) {
                        // dd($authUser);
                        // echo "true";exit;
                        // No need to call Auth::login() if the user is already logged in.
                        // Just return the response and continue with the request.
                        return $next($request);
                    } else {
                        // dd($authUser);
                        // echo "false";exit;
                        // Auth::logout();
                        Notification::make()
                            ->title('Your account is deactivated!')
                            ->danger()
                            ->send();
                    }
                } else {
                    Auth::logout();
                    Notification::make()
                        ->title('You do not have permission to log in to this account!')
                        ->danger()
                        ->send();
                }
            }
        }
        return $next($request);
    }
}

**Reply by mohandtharsha** [reply]
I am getting the below error,

Filament\FilamentManager::getUserAvatarUrl(): Argument #1 ($user) must be of type Illuminate\Database\Eloquent\Model|Illuminate\Contracts\Auth\Authenticatable, null given, called in C:\xampp\htdocs\EldersTalentNew\storage\framework\views\96bb8307b3bfe33a8c5d0701809c42c5.php on line 22

**Reply by mohandtharsha** [reply]
not added

**Reply by epcgrs** [reply]
Have the same problem here, 
I have a custom path to filament: /admin
I note the error with getUserAvatarUrl() are showing if I not logged in on route /admin
and if I access /admin/login its working. 

So I think this is a problem with a middleware to redirect if is not authenticated

**Reply by darshan45672** [reply]
you must register the middleware in the admin panel provider. if its not register there than that middleware wont work on filament panel.

**Reply by epcgrs** [reply]
@darshan45672 I found the problem, in my case, I had not implemented the FilamentUser interface in the model. Sorry and Thank you
